src/Std/Sync/Channel.lean

```diff
@@ -423,13 +543,75 @@ private partial def recv (ch : Bounded α) : BaseIO (Task (Option α)) := do
       return .pure none
     else
       let promise ← IO.Promise.new
-      modify fun st => { st with consumers := st.consumers.enqueue promise }
+      modify fun st => { st with consumers := st.consumers.enqueue ⟨promise, none⟩ }
       BaseIO.bindTask promise.result? fun res => do
         if res.getD false then
           Bounded.recv ch
         else
           return .pure none
 
+@[inline]
+private def recvReady' [Monad m] [MonadLiftT (ST IO.RealWorld) m] :
+    AtomicT (Bounded.State α) m Bool := do
+  let st ← get
+  return st.bufCount != 0 || st.closed
+
+open Internal.IO.Async in
+private partial def recvSelector (ch : Bounded α) : Selector (Option α) :=
+  {
+    tryFn := do
+      ch.state.atomically do
+        if ← recvReady' then
+          let val ← tryRecv'
+          return some val
+        else
+          return none
+
+    registerFn := registerAux ch
+
+    unregisterFn := do
+      ch.state.atomically do
+        let st ← get
+        let consumers ← st.consumers.filterM fun c => do
+          match c.waiter with
+          | some waiter => return !(← waiter.checkFinished)
+          | none => return true
+
+        set { st with consumers }
+  }
+where
+  registerAux (ch : Bounded α) (waiter : Waiter (Option α)) : IO Unit := do
+    ch.state.atomically do
+      -- We did drop the lock between `tryFn` and now so maybe ready?
+      if ← recvReady' then
+        let lose := return ()
+        let win promise := do
+          -- We know we are ready so the value by this is fine
+          promise.resolve (.ok (← tryRecv'))
+
+        waiter.race lose win
+      else
+        let promise ← IO.Promise.new
+        modify fun st => { st with consumers := st.consumers.enqueue ⟨promise, some waiter⟩ }
+
+        IO.chainTask promise.result? fun res? => do
+          match res? with
+          | none => return ()
+          | some res =>
+            if res then
+              registerAux ch waiter
+            else
+              -- if we lose we must trigger the next promise (if available) to avoid deadlocking
+              let lose := do
+                ch.state.atomically do
+                  let st ← get
+                  if let some (consumer, consumers) := st.consumers.dequeue? then
+                    consumer.resolve true
```

Comment:
I don't understand what's going on here. Why `true`? Isn't the channel closing in this branch or am I completely misunderstanding something?

Reply:
No, this is the branch that we take when we loose the race to fulfill the select. E.g. we have two channels and a message arrives on both at the same time. Only one of them can win the `waiter.race` call below, the other will go into the loosing branch. As it goes into the loosing branch it is not allowed to fetch the data from the channel so if there is a next consumer available its promise should be resolved such that the piece of data isn't left in the channel until the next send happens.

Reply:
Oops this code ended up in the wrong branch yes /o\